File: frontend/src/services/applicationService.js
import { apiRequest } from "./api";

export const createApplication = async (applicationData) => {
  return apiRequest("/api/applications", {
    method: "POST",
    body: JSON.stringify(applicationData),
  });
};

export const getAllApplications = async (
  pageNo = 0,
  pageSize = 10,
  sortBy = "id",
  sortDir = "asc",
) => {
  const params = new URLSearchParams({
    pageNo,
    pageSize,
    sortBy,
    sortDir,
  });

  return apiRequest(`/api/applications?${params}`);
};

export const getApplicationById = async (id) => {
  return apiRequest(`/api/applications/${id}`);
};

export const updateApplication = async (id, applicationData) => {
  return apiRequest(`/api/applications/${id}`, {
    method: "PUT",
    body: JSON.stringify(applicationData),
  });
};

export const deleteApplication = async (id) => {
  return apiRequest(`/api/applications/${id}`, {
    method: "DELETE",
  });
};

export const updateApplicationStatus = async (id, status) => {
  return apiRequest(`/api/applications/${id}/${encodeURIComponent(status)}`, {
    method: "PUT",
  });
};

export const getApplicationsByCandidateId = async (
  candidateId,
  pageNo = 0,
  pageSize = 10,
) => {
  const params = new URLSearchParams({
    pageNo,
    pageSize,
  });

  return apiRequest(`/api/applications/candidate/${candidateId}?${params}`);
};

export const getApplicationsByJobId = async (
  jobId,
  pageNo = 0,
  pageSize = 10,
) => {
  const params = new URLSearchParams({
    pageNo,
    pageSize,
  });

  return apiRequest(`/api/applications/job/${jobId}?${params}`);
};

export const getApplicationsByStatus = async (
  status,
  pageNo = 0,
  pageSize = 10,
) => {
  const params = new URLSearchParams({
    pageNo,
    pageSize,
  });

  return apiRequest(
    `/api/applications/status/${encodeURIComponent(status)}?${params}`,
  );
};

export const getApplicationsByCandidateName = async (
  candidateName,
  pageNo = 0,
  pageSize = 10,
) => {
  const params = new URLSearchParams({
    pageNo,
    pageSize,
  });

  return apiRequest(
    `/api/applications/candidateName/${encodeURIComponent(candidateName)}?${params}`,
  );
};

export const getApplicationsByJobTitle = async (
  jobTitle,
  pageNo = 0,
  pageSize = 10,
) => {
  const params = new URLSearchParams({
    pageNo,
    pageSize,
  });

  return apiRequest(
    `/api/applications/jobTitle/${encodeURIComponent(jobTitle)}?${params}`,
  );
};

File: backend/src/main/java/com/amritan/backend/service/ApplicationService.java
package com.amritan.backend.service;

import com.amritan.backend.dto.ApplicationDto;
import com.amritan.backend.dto.PageResponse;
import com.amritan.backend.enums.ApplicationStatus;

public interface ApplicationService {

	ApplicationDto createApplication(ApplicationDto dto);
	
	ApplicationDto applyForJob(Long jobId);

	PageResponse<ApplicationDto> getAllApplications(int pageNo, int pageSize,
			String sortBy, String sortDir);
	
	PageResponse<ApplicationDto> getMyApplications(
	        int pageNo,
	        int pageSize
	);

	ApplicationDto getApplicationById(Long id);
	
	ApplicationDto updateStatus(Long id, ApplicationStatus status);
	
	ApplicationDto updateApplication(Long id, ApplicationDto dto);
	
	void deleteApplication(Long id);

	PageResponse<ApplicationDto> getApplicationsByCandidateId(Long candidateId, int pageNo, int pageSize);

    PageResponse<ApplicationDto> getApplicationsByJobId(Long jobId, int pageNo, int pageSize);
    
    PageResponse<ApplicationDto> getApplicationByStatus(ApplicationStatus status, int pageNo, int pageSize);
    
    PageResponse<ApplicationDto> getApplicationByCandidateName(String keyword, int pageNo, int pageSize);
    
    PageResponse<ApplicationDto> getApplicationByJobTitle(String keyword, int pageNo, int pageSize);    
}
